commander.go

```diff
@@ -0,0 +1,295 @@
+	// return immediately
+	if len(c.cmds) < 1 {
+		return
+	}
+
+	if c.flags.NArg() < 1 {
+		c.flags.Usage()
+		os.Exit(1)
+	}
+
+	name := c.flags.Arg(0)
+
+	if cont, ok := c.cmds[name]; ok { //this is an existing command
+
+		// Init it
+		c.matchingFlags = cont.command.Flags(flag.NewFlagSet(name, flag.ExitOnError))
+		// always append a -h option to print "help"
+		c.flagHelp = c.matchingFlags.Bool("h", false, "")
+		c.matchingFlags.Parse(c.flags.Args()[1:])
+		c.matchingCmd = cont
+
+		// recursive case: if it's also a commander (it has subcommands)
+		if cmdr, ok := cont.command.(Commander); ok {
+			cmdr.SetName(c.name + " " + name)
+			//we had to split setName, and Parse because:
+
+			// checking for required flags might target subCommandUsage (that needs the name to be set)
+			// but this check need to be made before parsing the subcommand
+		}
+
+		// Check for required flags.
+		flagMap := make(map[string]bool)
+		for _, flagName := range cont.requiredFlags {
+			flagMap[flagName] = true
+		}
+		c.matchingFlags.Visit(func(f *flag.Flag) {
+			delete(flagMap, f.Name)
+		})
+		if len(flagMap) > 0 { // missed a required flag
+			c.subcommandUsage(c.matchingCmd)
+			os.Exit(1)
+		}
+
+		// ok this one is good. Now, if this "Cmd" is also a Commander go on parsing
+		if cmdr, ok := cont.command.(Commander); ok {
+			cmdr.Parse() // fs has been parsed
+		}
+
+	} else {
+		c.flags.Usage()
+		os.Exit(1)
+	}
+}
+
+//Implement the Cmd .Run method so that a Commander is almsot a Cmd too.
+// the args is ignored.
+func (c *commander) Run(args []string) {
+	c.Exec()
+}
+
+// Exec the subcommand's runnable. If there is no subcommand
+// registered, it silently returns.
+//args is totally ignored and kept to be
+func (c *commander) Exec() {
+	if c.matchingCmd != nil {
+		if *c.flagHelp {
+			c.subcommandUsage(c.matchingCmd)
+			return
+		}
+		c.matchingCmd.command.Run(c.matchingFlags.Args())
+	}
+}
+
+// Prints the usage.
+func (c *commander) Usage() {
+	name := c.name
+	if len(c.cmds) == 0 {
+		// no subcommands
+		fmt.Fprintf(os.Stderr, "Usage of %s:\n", name)
+		c.flags.PrintDefaults()
+		return
+	}
+
+	fmt.Fprintf(os.Stderr, "Usage: %s <command>\n\n", name)
+	fmt.Fprintf(os.Stderr, "where <command> is one of:\n")
+
+	//Sort conts by name
+	conts := make([]*cmdCont, 0, len(c.cmds))
+	for _, cont := range c.cmds {
+		conts = append(conts, cont)
+	}
+	sort.Stable(byName(conts))
+
+	for _, cont := range conts {
+		fmt.Fprintf(os.Stderr, "  %-15s %s\n", cont.name, cont.desc)
+	}
+
+	if numOfGlobalFlags(c.flags) > 0 {
+		fmt.Fprintf(os.Stderr, "\navailable flags:\n")
+		c.flags.PrintDefaults()
+	}
+	fmt.Fprintf(os.Stderr, "\n%s <command> -h for subcommand help\n", name)
+}
+
+func (c *commander) subcommandUsage(cont *cmdCont) {
+	name := c.name
+	fmt.Fprintf(os.Stderr, "  %s %-15s %s\n", name, cont.name, cont.desc)
+
+	// should only output sub command flags, ignore h flag.
+	fs := cont.command.Flags(flag.NewFlagSet(cont.name, flag.ContinueOnError))
```

Comment:
btw I fixed a small error here: it was fs:= matchingCmd.Flags   instead of fs:= cont.Flags 

This had no effect since subCommandUsage was always called with matchingCmd anyway.